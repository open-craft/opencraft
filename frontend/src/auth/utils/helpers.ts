import { V2Api } from 'global/api';

export const checkAuthAndRefreshToken = async () => {
  const refreshToken = localStorage.getItem('token_refresh');
  let response = null;

  if (refreshToken) {
    try {
      response = await V2Api.authRefreshCreate({
        data: { refresh: refreshToken }
      });
    } catch (e) {
      return false;
    }
  }

  if (response) {
    localStorage.setItem('token_access', response.access);
    return true;
  }

  return false;
};
