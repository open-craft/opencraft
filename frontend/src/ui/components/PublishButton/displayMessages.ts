const messages = {
  deploy: {
    defaultMessage: 'Publish changes',
    description: 'Deployment/Publish button text.'
  },
  notificationText: {
    defaultMessage: '{notificationNumber}',
    description: ''
  }
};

export default messages;
