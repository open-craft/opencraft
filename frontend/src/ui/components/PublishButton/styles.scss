@import '~styles/theme';

.notification-layer {
    position: relative;
    margin-top: -25px;
    left: 135px;
    z-index:2;
    background-color:#dd0000;
    height: 24px;
    width: 24px;
    border-radius: 100%;
    overflow:  visible;
    border-style: solid;
    border-color: white;
    border-width: 2px;
    box-shadow: 0 2px 4px 0 rgba(15, 31, 36, 0.5);
  }
  .notification-layer p {
    display: inline;
    text-align: center;
    vertical-align: middle;
    font-size: 14px;
    font-weight: normal;
    font-stretch: normal;
    font-style: normal;
    color:#ffffff;
    letter-spacing: -0.5px;
    padding-top: 6px;
  }

  .button-text-layer {
    z-index:1;
    margin-top: 1px;
    margin-left: -29px;
    font-size: 14px;
    font-weight: normal;
    font-stretch: normal;
    font-style: normal;
    font-family: $chivo-font;
    line-height: normal;
    letter-spacing: 1px;
    width: 180px;
  }

  .btn-disabled {
    background-color:rgb(255, 255, 255) !important;
    border-color: grey !important;
    color:grey !important;
    opacity: 0.3;
    width: 181px;
    height: 49px;
  }

  .btn-enabled {
    background-color:#ffffff !important;
    border: solid 1px #1e4c59 !important;
    color: #1e4c59 !important;
    opacity: 1 !important;
    width: 181px;
    height: 49px;
  }

