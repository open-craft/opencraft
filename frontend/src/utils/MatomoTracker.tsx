import React, { ReactNode } from 'react';
import {
  createInstance,
  MatomoProvider,
  useMatomo
} from '@datapunt/matomo-tracker-react';
import {
  MATOMO_ALIAS_DOMAIN,
  MATOMO_BASE_URL,
  MATOMO_MY_DOMAIN,
  MATOMO_SITE_ID
} from '../global/constants';

interface MatomoTrackerProps {
  children: ReactNode;
}

interface MatomoAdditionalConfiguration {
  [propName: string]: any;
}

interface MatomoConfiguration {
  urlBase: string;
  siteId: number;
  configurations?: MatomoAdditionalConfiguration;
}

interface MatomoUserIdTrackerProps {
  userId: string;
}

/*
 * This wrapper component initializes the Matomo tracking by wrapping the children
 * elements, if the tracking is enabled by providing a valid value for the 'MATOMO_BASE_URL'
 * constant. It allows tracking page views, events and the user id, when available.
 *
 * The 'MATOMO_ALIAS_DOMAIN' constant allows the cross-domain linking of another site with
 * this site so that they are treated as the same site in Matomo.
 */
export const MatomoTracker: React.FC<MatomoTrackerProps> = (
  props: MatomoTrackerProps
) => {
  let element = props.children;
  if (MATOMO_BASE_URL) {
    const matomoConfiguration: MatomoConfiguration = {
      urlBase: MATOMO_BASE_URL,
      siteId: parseInt(MATOMO_SITE_ID, 10)
    };
    if (MATOMO_ALIAS_DOMAIN) {
      const myDomain = `*.${MATOMO_MY_DOMAIN}`;
      const aliasDomain = `*.${MATOMO_ALIAS_DOMAIN}`;
      matomoConfiguration.configurations = {
        enableCrossDomainLinking: true,
        setDomains: [myDomain, aliasDomain]
      };
    }
    element = (
      <MatomoProvider value={createInstance(matomoConfiguration)}>
        {props.children}
      </MatomoProvider>
    );
  }
  return <>{element}</>;
};

/**
 * This is a component passes the provided user id to Matomo if Matomo tracking is enabled.
 * The 'pushInstruction' function exposed by the 'useMatomo()' hook is used to do this.
 *
 * This component is needed as all the components in the registration process are classes and hence
 * the 'yseMatomo' hook cannot be directly used in those components.
 */
export const MatomoUserIdTracker: React.FC<MatomoUserIdTrackerProps> = (
  props: MatomoUserIdTrackerProps
) => {
  const { pushInstruction } = useMatomo();
  if (MATOMO_BASE_URL && props.userId) {
    pushInstruction('setUserId', props.userId);
  }
  return null;
};
