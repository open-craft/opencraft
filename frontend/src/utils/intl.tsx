import React from "react";
import { FormattedMessage, MessageDescriptor } from "react-intl";

type MessageMap = { [k: string]: Omit<MessageDescriptor, 'id'> };

interface Props<T, K extends keyof T> {
    id: K;
    messages: T;
    values?: any;
}


export const WrappedMessage = (props: Props<MessageMap, string>) => {
    const {messages, id, ...values} = props;
    return (
        <FormattedMessage  {...messages[id]} id={id} {...values}/>
    );
};
