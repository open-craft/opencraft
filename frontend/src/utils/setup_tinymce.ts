import 'tinymce/tinymce';
import 'tinymce/themes/silver';
import 'tinymce/skins/ui/oxide/skin.min.css';
import 'tinymce/skins/ui/oxide/content.min.css';
import 'tinymce/plugins/advlist';
import 'tinymce/plugins/autolink';
import 'tinymce/plugins/lists';
import 'tinymce/plugins/link';
import 'tinymce/plugins/image';
import 'tinymce/plugins/charmap';
import 'tinymce/plugins/print';
import 'tinymce/plugins/preview';
import 'tinymce/plugins/anchor';
import 'tinymce/plugins/searchreplace';
import 'tinymce/plugins/visualblocks';
import 'tinymce/plugins/code';
import 'tinymce/plugins/fullscreen';
import 'tinymce/plugins/insertdatetime';
import 'tinymce/plugins/media';
import 'tinymce/plugins/table';
import 'tinymce/plugins/paste';
import 'tinymce/plugins/help';
import 'tinymce/plugins/wordcount';
