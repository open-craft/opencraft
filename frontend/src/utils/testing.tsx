import { createMemoryHistory } from 'history';
import * as React from 'react';
import { IntlProvider } from 'react-intl';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router';
import * as renderer from 'react-test-renderer';
import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import { createRootReducer } from '../global/reducers';

export const setupComponentForTesting = (
  reactContent: JSX.Element,
  storeContents = {}
) => {
  const middleware = applyMiddleware(thunk);

  const store = createStore(
    createRootReducer(createMemoryHistory()),
    storeContents,
    middleware
  );

  return renderer.create(
    <IntlProvider textComponent={React.Fragment} locale="en">
      <Provider store={store}>
        <MemoryRouter>{reactContent}</MemoryRouter>
      </Provider>
    </IntlProvider>
  );
};
