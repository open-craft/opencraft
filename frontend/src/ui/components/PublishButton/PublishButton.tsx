import * as React from 'react';
import { WrappedMessage } from 'utils/intl';
import { Button } from 'react-bootstrap';
import messages from './displayMessages';
import './styles.scss';

interface Props {
  deploymentDisabled: boolean;
  undeployedChanges: number;
  deploymentHandler: Function;
}

export const PublishButton: React.FC<Props> = ({
  deploymentDisabled,
  undeployedChanges,
  deploymentHandler
}: Props) => {
  const notificationNumber = undeployedChanges <= 9 ? undeployedChanges : '9+';

  return (
    <Button
      className={
        deploymentDisabled
          ? 'float-right btn-disabled'
          : 'float-right btn-enabled'
      }
      variant="primary"
      size="lg"
      disabled={deploymentDisabled}
      onClick={() => {
        deploymentHandler();
      }}
    >
      {undeployedChanges !== 0 && (
        <div className="notification-layer">
          <p>
            <WrappedMessage
              id="notificationText"
              messages={messages}
              values={{ notificationNumber }}
            />
          </p>
        </div>
      )}
      <div className="button-text-layer">
        <WrappedMessage id="deploy" messages={messages} />
      </div>
    </Button>
  );
};
