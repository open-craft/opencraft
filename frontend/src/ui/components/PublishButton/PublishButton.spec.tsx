import React from 'react';
import { setupComponentForTesting } from "utils/testing";
import { PublishButton } from "./PublishButton"
/**
 * Returns the PublishButton's test props.
 *
 * @param {Object}  component - The root component.
 * @return {Array<String>} The specified props as an array of strings
 */
function getTestProps(publishBtn: Object): Array<String> {
  const disabled = publishBtn.props["disabled"]
  const notificationCount = publishBtn.children[0].props.className === "notification-layer"
    ? publishBtn.children[0].children[0].children[0]
    : null
  return [disabled, notificationCount]
}

const singleDigitChanges = Array.from('x'.repeat(5)).length
const doubleDigitChanges = Array.from('x'.repeat(15)).length

it('renders without crashing', () => {
    const tree = setupComponentForTesting(<PublishButton />).toJSON();
    expect(tree).toMatchSnapshot();
});

describe("PublishButton renders as disabled when", function() {
  it('pending changes', () => {
      const publishBtn = setupComponentForTesting(
      <PublishButton
        undeployedChanges={singleDigitChanges}
        deploymentDisabled={true}
        onClickWrapper={function() { return false; }}
      />
      ).toJSON();
      expect(getTestProps(publishBtn)).toEqual([true, "5"])
  });

  it('no pending changes', () => {
    const publishBtn = setupComponentForTesting(
    <PublishButton
    undeployedChanges={[].length}
    deploymentDisabled={true}
    onClickWrapper={function() { return false; }}
    />
    ).toJSON();
    expect(getTestProps(publishBtn)).toEqual([true, null])
  });
});

describe("PublishButton renders as enabled when", function() {
  it('single-digit pending changes', () => {
    const publishBtn = setupComponentForTesting(
    <PublishButton
      undeployedChanges={singleDigitChanges}
      deploymentDisabled={false}
      onClickWrapper={function() { return false; }}
    />
    ).toJSON();
    expect(getTestProps(publishBtn)).toEqual([false, "5"])
  });
  it('double-digit pending changes', () => {
    const publishBtn = setupComponentForTesting(
    <PublishButton
      undeployedChanges={doubleDigitChanges}
      deploymentDisabled={false}
      onClickWrapper={function() { return false; }}
    />
    ).toJSON();
    expect(getTestProps(publishBtn)).toEqual([false, "9+"])
  });
})
