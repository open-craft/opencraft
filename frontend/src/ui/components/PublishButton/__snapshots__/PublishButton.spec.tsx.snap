// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<button
  className="float-right btn-enabled btn btn-primary btn-lg"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  <div
    className="notification-layer"
  >
    <p>
      9+
    </p>
  </div>
  <div
    className="button-text-layer"
  >
    Publish changes
  </div>
</button>
`;
